jax/experimental/checkify.py

```diff
@@ -0,0 +1,269 @@
+
+@dataclass(frozen=True)
+class Error:
+  err: Union[bool, core.Tracer]
+  code: Union[int, core.Tracer]
+  msgs: Dict[int, str]
+
+  def get(self) -> Optional[str]:
+    assert np.shape(self.err) == np.shape(self.code)
+    if np.size(self.err) == 1:
+      if self.err:
+        return self.msgs[int(self.code)]
+    else:
+      return '\n'.join(f'at mapped index {", ".join(map(str, idx))}: '  # type: ignore
+                       f'{self.msgs[int(self.code[idx])]}'              # type: ignore
+                       for idx, e in np.ndenumerate(self.err) if e) or None
+    return None
+
+register_pytree_node(Error,
+                     lambda e: ((e.err, e.code), tuple(sorted(e.msgs.items()))),
+                     lambda msgs, data: Error(*data, dict(msgs)))  # type: ignore
+
+init_error = Error(False, 0, {})
+next_code = it.count(1).__next__  # globally unique ids, could be uuid4
+
+
+Bool = Union[bool, core.Tracer]
+Int = Union[int, core.Tracer]
+
+def assert_func(error: Error, pred: Bool, msg: str) -> Error:
+  code = next_code()
+  out_err = error.err | jnp.logical_not(pred)
+  out_code = lax.select(error.err, error.code, code)
+  return Error(out_err, out_code, {code: msg, **error.msgs})
+
+## Checkify transformation for plumbing functional error values.
+
+class ErrorTrace(core.Trace):
+  pure = lift = sublift = lambda self, val: ErrorTracer(self, val)
+
+  def process_primitive(self, primitive, tracers, params):
+    in_vals = [t.val for t in tracers]
+    rule = error_checks.get(primitive)
+    if rule:
+      out, self.main.error = rule(self.main.error, *in_vals, **params)
+    else:
+      out = primitive.bind(*in_vals, **params)
+    if primitive.multiple_results:
+      return [ErrorTracer(self, x) for x in out]
+    else:
+      return ErrorTracer(self, out)
+
+  def process_call(self, primitive, f, tracers, params):
+    in_vals = [t.val for t in tracers]
+    e = popattr(self.main, 'error')
+    f, msgs = check_errors_subtrace(f, self.main, tuple(e.msgs.items()))
+    params_ = dict(params, donated_invars=(False, False, *params['donated_invars']))
+    err, code, *out_vals = primitive.bind(f, e.err, e.code, *in_vals, **params_)
+    setnewattr(self.main, 'error', Error(err, code, msgs()))
+    return [ErrorTracer(self, x) for x in out_vals]
+
+  def process_map(self, primitive, f, tracers, params):
+    in_vals = [t.val for t in tracers]
+    e = popattr(self.main, 'error')
+    f, msgs = check_errors_subtrace(f, self.main, tuple(e.msgs.items()))
+
+    @as_hashable_function(closure=params['out_axes_thunk'])
+    def new_out_axes_thunk():
+      return (0, 0, *params['out_axes_thunk']())
+
+    params_ = dict(params, in_axes=(None, None, *params['in_axes']),
+                   out_axes_thunk=new_out_axes_thunk,
+                   donated_invars=(False, False, *params['donated_invars']))
+    errs, codes, *outs = primitive.bind(f, e.err, e.code, *in_vals, **params_)
+    err, code = _reduce_any_error(errs, codes)
```

Comment:
I don't understand what _reduce_any_error does: sorts the boolean errors and returns the last one? So it would return a True error if there is one, but if there's multiple it would return an arbitrary one?

Reply:
Will it not return the code with the highest index because it returns the last element? Or am I thinking in the wrong dimension?

And is this "code with the lowest index" the "first" error which occurred in all mapped computations, or the error which occurred at the lowest mapped element? I think it's the latter, in which case I guess it doesn't matter which of the True codes we return (if we're only going to return one).